openedx/features/course_experience/views/course_outline.py

```diff
@@ -23,15 +26,44 @@ def render_to_fragment(self, request, course_id=None, page_context=None, **kwarg
         """
         course_key = CourseKey.from_string(course_id)
         course_overview = get_course_overview_with_access(request.user, 'load', course_key, check_if_enrolled=True)
-
         course_block_tree = get_course_outline_block_tree(request, course_id)
+
         if not course_block_tree:
             return None
 
+        content_milestones = self.get_content_milestones(request, course_key)
+
         context = {
             'csrf': csrf(request)['csrf_token'],
             'course': course_overview,
             'blocks': course_block_tree,
+            'milestones': content_milestones
         }
         html = render_to_string('course_experience/course-outline-fragment.html', context)
         return Fragment(html)
+
+    def get_content_milestones(self, request, course_key):
+        """
+        Returns dict of subsections with prerequisites and whether the prerequisite has been completed or not
+        """
+        course_content_milestones = {}
+
+        all_course_prereqs = get_course_content_milestones(
+            course_id=course_key,
+            content_id=None,
+            relationship='requires',
+            user_id=None)
+
+        for milestone in all_course_prereqs:
+            course_content_milestones[milestone['content_id']] = {'completed_prereqs': True}
+            
+        unfulfilled_prereqs = get_course_content_milestones(
+            course_id=course_key,
+            content_id=None,
+            relationship='requires',
+            user_id=request.user.id)
+
+        for milestone in unfulfilled_prereqs:
+            course_content_milestones[milestone['content_id']]['completed_prereqs'] = False
```

Comment:
Is there any chance that a milestone appears in `unfulfilled_prereqs` but wasn't present in `all_course_prereqs`?  Due to data inconsistency, for example?

Suggestion: One way you can safeguard against this is to compute `all_course_prereqs` and `unfulfilled_prereqs` and then only `course_content_milestones`.  For example:
```
all_course_prereqs = get_course_content_milestones(course_key)
content_ids_of_unfulfilled_prereqs = [
    milestone['content_id']
    for milestone in get_course_content_milestones(course_key, user_id=request.user.id)
]
course_content_milestones = {
    milestone['content_id']: {
        'completed_prereqs': milestone['content_id'] not in content_ids_of_unfulfilled_prereqs
    }
    for milestone in all_course_prereqs
}
```

Reply:
I hadn't thought about that case - good point!